File: internal/transport/http/execution.go
package http

import (
	"Algolearn/internal/execution/manager"

	"github.com/gofiber/fiber/v3"
)

type ExecutionHandler struct {
	manager *manager.Manager
}

func NewExecutionHandler(m *manager.Manager) *ExecutionHandler {
	return &ExecutionHandler{
		manager: m,
	}
}

type ExecuteRequest struct {
	Algorithm string `json:"algorithm"`
	Data      []int  `json:"data"`
}

func (h *ExecutionHandler) Execute(c fiber.Ctx) error {
	var req ExecuteRequest

	if err := c.Bind().Body(&req); err != nil {
		return c.SendStatus(fiber.StatusBadRequest)
	}

	steps, err := h.manager.Execute(req.Algorithm, req.Data)
	if err != nil {
		return c.SendStatus(fiber.StatusBadRequest)
	}

	return c.JSON(fiber.Map{
		"steps": steps,
	})
}

File: frontend/src/api/execution.ts
import { http } from "./http";

export async function executeAlgorithm(algorithm_name: string, data: number[]) {
   return http('/api/execute', {
    method: "POST",
    body: JSON.stringify({
      algorithm: algorithm_name,
      data: data,
    })
   })
}

export function getExecution(id: string) {
  return http(`/api/execution/${id}`);
}